walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -64,6 +65,215 @@
+   */
+  public Collection<Event> getOptionalOnlyEvents(Collection<Event> events, MeetingRequest request) {
+    return getOptionalOnlyEventsInRange(TimeRange.WHOLE_DAY, events, request);
+  }
+
+  /**
+   * This method gets all the unavailability slots throughout a day.
+   * @param optionalOnlyEvents represents all the optional only events throughout the dat
+   * @param request is the original MeetingRequest object
+   * @return Collection<TimeRangeAndUnavailable> is the entire day broken into slots in which unavailability is constant.
+   */
+  public Collection<TimeRangeAndUnavailable> getAllUnavailability(Collection<Event> optionalOnlyEvents, MeetingRequest request) {
+    List<TimeRangeAndUnavailable> unavailability = new ArrayList<>();
+    PriorityQueue<EventAndTime> eventTimesPQ = new PriorityQueue<>(5, EventAndTime.ORDER);
+    
+    for (Event e : optionalOnlyEvents) {
+      eventTimesPQ.add(new EventAndTime(e, true));
+      eventTimesPQ.add(new EventAndTime(e, false));
+    }
+
+    int numUnavailable = 0;
+    if (eventTimesPQ.isEmpty()) return unavailability;
+    EventAndTime one = eventTimesPQ.poll();
+    if (one.getTime() != TimeRange.START_OF_DAY) {
+      unavailability.add(new TimeRangeAndUnavailable(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, one.getTime(), false), 0));
+    }
+    numUnavailable += one.getEvent().getAttendees().size();
+    while (!eventTimesPQ.isEmpty()) {
+      EventAndTime two = eventTimesPQ.poll();
+
+      //Is the range long enough to hold the requested meeting?
+      if (two.getTime() - one.getTime() >= request.getDuration()) {
+        unavailability.add(new TimeRangeAndUnavailable(TimeRange.fromStartEnd(one.getTime(), two.getTime(), false), numUnavailable));
+      } 
+
+      //If the second time starts a new event, add its attendees to numUnavailable.
+      if (two.isStart()) {
+        numUnavailable += two.getEvent().getAttendees().size();
+      }
+
+      //If the second time ends an event, remove its attendees from numUnavailable.
+      else {
+        numUnavailable -= two.getEvent().getAttendees().size();
+      }
+
+      one = two;
+    }
+    if (one.getTime() != TimeRange.END_OF_DAY) {
+      unavailability.add(new TimeRangeAndUnavailable(TimeRange.fromStartEnd(one.getTime(), TimeRange.END_OF_DAY, true), 0));
+    }
+
+    return unavailability;
+  }
+
+  /**
+   * This method retrieves a list of the optimal free times
+   * @param openRanges is the list of free times, based on mandatory attendees
+   * @param events is all the events throughout the day
+   * @param request is the original MeetingRequest object.
+   * @return Collection<TimeRangeAndUnavailable> is a list of the optimal timeranges, and their associated unavailability of optional attendees.
+   */
+  public Collection<TimeRangeAndUnavailable> optimalTimeAndUnavailables(Collection<TimeRange> openRanges, Collection<Event> events, MeetingRequest request) {
```

Comment:
This function is getting pretty big - looking overall at its structure, it looks like it's handling 4 major cases. Consider splitting each case into its own helper function?

Reply:
I think that's a great idea. This function sort of got away from me, and I think that's what led to some of my silly mistakes along the way. I'll refactor and clean it up, and let you know when it's fixed.